### PR TITLE
Revamp the droid calendar

### PR DESCRIPTION
**Pull request by daividssilverio**
## What's this?
This basically re-writes the current droid calendar implementation using Canvas.

### Relationships
Closes Nothing

## Why do we want this?
Because only I can work with the current implementation and I regret what I did back then.
All of that logic and boilerplate needed to maintain the custom implementation of the `CalendarLayoutManager` was overkill.

Please take a look at these files to understand why I think this PR is necessary before we add new features to the calendar.

```
Toggl.Droid/Adapters/Calendar/Anchor.cs
Toggl.Droid/Adapters/Calendar/AnchorData.cs
Toggl.Droid/Adapters/Calendar/CalendarAdapter.cs
Toggl.Droid/Adapters/Calendar/LayoutAnchorsCalculator.cs
Toggl.Droid/Views/CalendarRecyclerView.cs
Toggl.Droid/Views/Calendar/CalendarLayoutManager.cs
Toggl.Droid/Views/Calendar/CalendarLayoutManager.AnchorInfo.cs
Toggl.Droid/Views/Calendar/CalendarLayoutManager.AnchorSavedState.cs
Toggl.Droid/Views/Calendar/CalendarLayoutManager.LayoutState.cs
Toggl.Droid/Views/Calendar/CalendarLayoutManager.Recycler.cs
Toggl.Droid/Views/Calendar/CalendarLayoutManager.ScrollHelper.cs
Toggl.Droid/ViewHolders/AnchorViewHolder.cs
Toggl.Droid/ViewHolders/CalendarEntryViewHolder.cs
```

They ☝️ are too complex and too codependent, making it hard to maintain. 

## How is it done?
The changes here replace the need for all the infrastructure added to support the current droid calendar.

### Why not another way?
We could have done something similar using a `ScrollView`, a bunch of `FrameLayouts` and implementing similar gesture detectors/listeners and some sort of manual view recycling, but it seems easier to just use a canvas instead of going through all of that trouble.

Feel free to suggest something better/easier to maintain.

### Side effects
We don't have the nice automatic animations in the current implementation. They are implementable and I intend to add them later.

The current implementation doesn't add the elevation effect when clicking a time entry for the first time. I intend to add it later when implementing the #5871 since it makes sense to do it there.

This implementation needs some extra testing because it hasn't been released to users. It might eve be beneficial to release it in 2.4 or 2.5 🤔 .

## Review hints
Please take a look at the old files and how it used to work and take compare with the new implementation.
Extra 🍪s for any performance improvements that can be easily added.

## :squid: Permissions
~🤔 not sure if we should 🦑 it into develop or wait for the calendar-improv branch to be cut and then merge it there. @sofokli1 @heytherewill what do you think?~

~Draft because of uncertainty of what should be the target of this PR.~
🦑 it. It's safe to do so because the bindings are not done yet.

### CODE REVIEW

**Comment by daividssilverio** [comment]
patch to help with tests: [calendar-canvas-test.txt](https://github.com/toggl/mobileapp/files/3624351/calendar-canvas-test.txt)

File: Toggl.Droid/Views/Calendar/CalendarDayView.Background.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+using Color = Android.Graphics.Color;
+using System.Globalization;
+using System.Linq;
+using Toggl.Droid.Extensions;
+
+namespace Toggl.Droid.Views.Calendar
+{
+    public partial class CalendarDayView
+    {
+        private static readonly string twelveHoursFormat = Shared.Resources.TwelveHoursFormat;
```
Comment (heytherewill) [inline Toggl.Droid/Views/Calendar/CalendarDayView.Background.cs]: Why cache this? Just `using static Toggl.Shared.Resource` and then you can access the property directly

File: Toggl.Droid/Views/Calendar/CalendarDayView.Background.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        private float timeSlicesTopPadding;
+        private float verticalLineLeftMargin;
+        private float hoursDistanceFromTimeLine;
+        private ImmutableArray<string> hours = ImmutableArray<string>.Empty;
+        private ImmutableArray<float> timeLinesYs = ImmutableArray<float>.Empty;
+        private ImmutableArray<float> hoursYs = ImmutableArray<float>.Empty;
+
+        private readonly Paint hoursLabelPaint = new Paint(PaintFlags.AntiAlias)
+        {
+            Color = Color.ParseColor("#757575"),
```
Comment (heytherewill) [inline Toggl.Droid/Views/Calendar/CalendarDayView.Background.cs]: Not dark theme aware.

File: Toggl.Droid/Views/Calendar/CalendarDayView.Background.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            TextAlign = Paint.Align.Right
+        };
+
+        private readonly Paint linesPaint = new Paint(PaintFlags.AntiAlias)
+        {
+            Color = Color.ParseColor("#19000000")
```
Comment (heytherewill) [inline Toggl.Droid/Views/Calendar/CalendarDayView.Background.cs]: Not dark theme aware

File: Toggl.Droid/Views/Calendar/CalendarDayView.Background.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        }
+        
+        private ImmutableArray<float> createTimeLinesYPositions()
+            => Enumerable.Range(0, hoursPerDay)
+                .Select(line => line * hourHeight + timeSlicesTopPadding)
+                .ToImmutableArray();
+        
+        private ImmutableArray<string> createHours()
+        {
+            DateTime date = new DateTime();
```
Comment (heytherewill) [inline Toggl.Droid/Views/Calendar/CalendarDayView.Background.cs]: var

File: Toggl.Droid/Views/Calendar/CalendarDayView.Events.cs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        private int regularCalendarItemVerticalPadding;
+        private int regularCalendarItemHorizontalPadding;
+        private int shortCalendarItemVerticalPadding;
+        private int shortCalendarItemHorizontalPadding;
+        
+        private RectF eventRect = new RectF();
+        
+        private readonly Paint eventsPaint = new Paint(PaintFlags.AntiAlias)
+        {
+            Color = Color.LightGray
```
Comment (heytherewill) [inline Toggl.Droid/Views/Calendar/CalendarDayView.Events.cs]: Not dark mode aware

File: Toggl.Droid/Views/Calendar/CalendarDayView.Events.cs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        };
+
+        private readonly Paint textEventsPaint = new Paint(PaintFlags.AntiAlias)
+        {
+            Color = Color.White,
```
Comment (heytherewill) [inline Toggl.Droid/Views/Calendar/CalendarDayView.Events.cs]: Not dark mode aware

**Comment by sofokli1** [comment]
if this is functionally the same as the previous implementation (and you fix what will says) I think this should be merged in develop

**Comment by sofokli1** [comment]
will mean less conflicts when merging the calendar project back into develop

File: Toggl.Droid/Views/Calendar/CalendarDayView.Events.cs (hunk 5 above)
Comment (daividssilverio) [reply Toggl.Droid/Views/Calendar/CalendarDayView.Events.cs]: Oops, leftover from tests. Events have their own colours.

File: Toggl.Droid/Views/Calendar/CalendarDayView.Events.cs (hunk 6 above)
Comment (daividssilverio) [reply Toggl.Droid/Views/Calendar/CalendarDayView.Events.cs]: Same thing, this color is always set based on the type fot the calendar item type.

**Pull request merged**